pkg/planner/core/logical_plan_builder.go

```diff
@@ -4907,8 +4942,12 @@ func (b *PlanBuilder) checkRecursiveView(dbName pmodel.CIStr, tableName pmodel.C
 // qbNameMap4View maps the query block name to the view table lists.
 // viewHints group the view hints based on the view's query block name.
 func (b *PlanBuilder) BuildDataSourceFromView(ctx context.Context, dbName pmodel.CIStr, tableInfo *model.TableInfo, qbNameMap4View map[string][]ast.HintTable, viewHints map[string][]*ast.TableOptimizerHint) (base.LogicalPlan, error) {
-	viewDepth := b.ctx.GetSessionVars().StmtCtx.ViewDepth
-	b.ctx.GetSessionVars().StmtCtx.ViewDepth++
+	stmtCtx := b.ctx.GetSessionVars().StmtCtx
+	viewDepth := len(stmtCtx.SavedViews)
+	stmtCtx.SavedViews = append(stmtCtx.SavedViews, &ast.TableName{Schema: dbName, Name: tableInfo.Name})
```

Comment:
`SavedViews` is a stack that saves all views when traversing the AST. 
We depend on it to report precise error in `appendColumnsToVisitedInfo`